### PR TITLE
Bug fix in cpu and gpu conditioning

### PR DESCRIPTION
**Pull request by yjchoi1**
**Describe the PR**
* Bug fix in cpu and gpu conditioning.
* Include loss in training_state file

### CODE REVIEW

File: gns/train.py [hunk 1]
```diff
@@ -296,15 +296,17 @@ def train(rank, flags, world_size):
         for param in optimizer.param_groups:
           param['lr'] = lr_new
 
-        if rank == 0 or torch.device("cpu"):
+        if rank == 0 or rank == torch.device("cpu"):
```
Comment (kks32) [inline gns/train.py:299]: Why is `rank`  has a value of `torch.device('cpu')`?

**Review by kks32 — CHANGES_REQUESTED** [review]
This doesn't make sense, rank should not be a device type

File: gns/train.py [hunk 2]
```diff
@@ -25,14 +25,14 @@
     help='Train model, validation or rollout evaluation.')
 flags.DEFINE_integer('batch_size', 2, help='The batch size.')
 flags.DEFINE_float('noise_std', 6.7e-4, help='The std deviation of the noise.')
-flags.DEFINE_string('data_path', None, help='The dataset directory.')
-flags.DEFINE_string('model_path', 'models/', help=('The path for saving checkpoints of the model.'))
+flags.DEFINE_string('data_path', "/work2/08264/baagee/frontera/gns-mpm-data/gns-data/datasets/WaterDropSample/", help='The dataset directory.')
```
Comment (kks32) [inline gns/train.py]:
```suggestion
flags.DEFINE_string('data_path', None, help='The dataset directory.')
```

Don't hardcode paths

File: gns/train.py [hunk 3]
```diff
@@ -25,14 +25,14 @@
...
+flags.DEFINE_string('model_path', '/work2/08264/baagee/frontera/gns-mpm-data/gns-data/models/WaterDropSample/', help=('The path for saving checkpoints of the model.'))
 flags.DEFINE_string('output_path', 'rollouts/', help='The path for saving outputs (e.g. rollouts).')
-flags.DEFINE_string('model_file', None, help=('Model filename (.pt) to resume from. Can also use "latest" to default to newest file.'))
-flags.DEFINE_string('train_state_file', 'train_state.pt', help=('Train state filename (.pt) to resume from. Can also use "latest" to default to newest file.'))
+flags.DEFINE_string('model_file', "model-10.pt", help=('Model filename (.pt) to resume from. Can also use "latest" to default to newest file.'))
+flags.DEFINE_string('train_state_file', 'train_state-10.pt', help=('Train state filename (.pt) to resume from. Can also use "latest" to default to newest file.'))
 
 flags.DEFINE_integer('ntraining_steps', int(2E7), help='Number of training steps.')
-flags.DEFINE_integer('nsave_steps', int(5000), help='Number of steps at which to save the model.')
```
Comment (kks32) [inline gns/train.py:35]: Revert this back

File: gns/train.py [hunk 4]
```diff
@@ -172,26 +172,24 @@ def optimizer_to(optim, device):
           if subparam._grad is not None:
             subparam._grad.data = subparam._grad.data.to(device)
 
-def train(rank, flags, world_size):
+def train(rank, flags, world_size, device):
   """Train the model.
 
   Args:
     rank: local rank
     world_size: total number of ranks
```
Comment (kks32) [inline gns/train.py:180]: If you add it to a function, add it to args

**Review by kks32 — CHANGES_REQUESTED** [review]

**Comment by yjchoi1** [comment]
Rolled back the flags.

File: gns/train.py [hunk 5]
```diff
@@ -213,14 +211,18 @@ def train(rank, flags, world_size):
...
 
       # load train state
       train_state = torch.load(flags["model_path"] + flags["train_state_file"])
       # set optimizer state
-      optimizer = torch.optim.Adam(simulator.module.parameters())
+      optimizer = torch.optim.Adam(
+        simulator.module.parameters() if device == torch.device("cuda") else simulator.parameters())
       optimizer.load_state_dict(train_state["optimizer_state"])
-      optimizer_to(optimizer, rank)
+      optimizer_to(optimizer, rank if device == torch.device("cuda") else device)
```
[the remark this answers is not in the document]
Comment (yjchoi1) [reply gns/train.py:227]: It indeed takes `rank`, which is an integer, if `torch.device('cuda'), but it takes `torch.device('cpu')` if cpu. That's the reason for the previous change I made [here](https://github.com/geoelements/gns/pull/60#discussion_r1300735377). However, as you pointed out in the previous comment, making `rank` to be `torch.device('cpu') does not sound natural, I made `optimizer` to take `rank` if cuda and `torch.device('cpu)` for the code to look more natural.

File: gns/train.py [hunk 6]
```diff
@@ -246,24 +248,24 @@ def train(rank, flags, world_size):
...
+        particle_type.to(rank if device == torch.device("cuda") else device)
+        n_particles_per_example.to(rank if device == torch.device("cuda") else device)
+        labels.to(rank if device == torch.device("cuda") else device)
 
         # TODO (jpv): Move noise addition to data_loader
         # Sample the noise to add to the inputs to the model during training.
-        sampled_noise = noise_utils.get_random_walk_noise_for_position_sequence(position, noise_std_last_step=flags["noise_std"]).to(rank)
-        non_kinematic_mask = (particle_type != KINEMATIC_PARTICLE_ID).clone().detach().to(rank)
+        sampled_noise = noise_utils.get_random_walk_noise_for_position_sequence(position, noise_std_last_step=flags["noise_std"]).to(rank if device == torch.device("cuda") else device)
+        non_kinematic_mask = (particle_type != KINEMATIC_PARTICLE_ID).clone().detach().to(rank if device == torch.device("cuda") else device)
```
[the remark this answers is not in the document]
Comment (yjchoi1) [reply gns/train.py:266]: To avoid repeating, the best way I can come up with is to predefine the rank as something like the following, since the rank takes integer if cuda else `torch.device('cpu')`
```
if torch.device('cpu'):
  rank = torch.device("cpu")
else:
  rank = 0   # If the primary device need to gather gradients.
```

File: gns/train.py [hunk 7]
```diff
@@ -178,11 +178,13 @@ def train(rank, flags, world_size, device):
   Args:
     rank: local rank
     world_size: total number of ranks
+    device: torch device type
   """
   if device == torch.device("cuda"):
     distribute.setup(rank, world_size, device)
 
   metadata = reading_utils.read_metadata(flags["data_path"])
+  metadata = metadata["train"]
```
Comment (kks32) [inline gns/train.py]: You are overwriting `metadata?` why not just concatenate and get the relevant data?

**Review by kks32 — CHANGES_REQUESTED** [review]

File: gns/train.py [hunk 8]
```diff
@@ -246,24 +249,24 @@ def train(rank, flags, world_size):
...
+      if device == torch.device("cuda"):
         torch.distributed.barrier()
       else:
         pass
       for ((position, particle_type, n_particles_per_example), labels) in dl:
-        position.to(rank)
-        particle_type.to(rank)
-        n_particles_per_example.to(rank)
-        labels.to(rank)
+        position.to(rank if device == torch.device("cuda") else device)
```
Comment (kks32) [inline gns/train.py:257]: Have a variable and use it here...don't have to check at every instance.

**Review by kks32 — CHANGES_REQUESTED** [review]

**Review by kks32 — APPROVED** [review]
